(dev/core#4147) Add test and fix for various flaky tokens

totten opened the pull request:
Overview
----------------------------------------

The discussion of [dev/core#4147](https://lab.civicrm.org/dev/core/-/issues/4147) highlighted that the token `{contact.address_primary.state_province_id:abbr}` behaves inconsistently -- in a way that resembles the previous [dev/core#4109](https://lab.civicrm.org/dev/core/-/issues/4109). In both cases, we have tokens that only work if they are used _in combination_; they fail individually.

cc @eileenmcnaughton @agileware-justin 

Before
----------------------------------------

Test available tokens in combination.

After
----------------------------------------

Test available tokens in combination and one-by-one.

Comments
----------------------------------------

Running locally, these sections stand out:

```diff
 contact.address_name:The white house\n
 contact.address_primary.city:New York\n
-contact.address_primary.country_id.region_id:name:America South, Central, North and Caribbean\n
+EXCEPTION:Undefined index: address_primary.country_id.region_id\n
 contact.address_primary.country_id:label:United States\n
 contact.address_primary.county_id:label:\n

 contact.address_primary.postal_code:90210\n
 contact.address_primary.postal_code_suffix:4578\n
-contact.address_primary.state_province_id:abbr:TX\n
+contact.address_primary.state_province_id:abbr:\n
 contact.address_primary.street_address:Street Address\n
 contact.address_primary.street_name:Main St\n

 contact.postal_greeting_display:Dear Robert\n
 contact.preferred_communication_method:label:Phone\n
-contact.preferred_communication_method:Phone\n
+EXCEPTION:Illegal offset type\n
 contact.preferred_language:label:French (Canada)\n
 contact.preferred_language:fr_CA\n
```

totten committed:
(REF) MessageTemplateTest - Extract method useFrozenTime()

totten committed:
(dev/core#4147) Add scenario testTokensIndividually

eileenmcnaughton commented:
Nice

totten commented on tests/phpunit/CRM/Core/BAO/MessageTemplateTest.php:
```diff
@@ -485,6 +485,52 @@ public function testContactTokens(): void {
+    // Evaluate all these token lines
+    $oneByOne = array_map($render, $tokenLines);
+    $allAtOnce = $render(implode("\n", $tokenLines));
+    $this->assertEquals($allAtOnce, implode("\n", $oneByOne));
```
This is the heart of the test - comparing `$oneByOne` with `$allAtOnce`.

totten commented on tests/phpunit/CRM/Core/BAO/MessageTemplateTest.php:
```diff
@@ -485,6 +485,52 @@ public function testContactTokens(): void {
+    foreach (array_keys($tokenData) as $key) {
+      $tokenLines[] .= "contact.$key:{contact.$key}";
+    }
+    $tokenLines = array_unique($tokenLines);
```
In reading `testContactTokens()`, I got a bit confused because there were a few distinct authorities providing token-lists, and they don't all agree. This takes all of them (i.e the union of "advertised tokens" and "old tokens").

eileenmcnaughton replied:
Yeah advertised tokens is a subset of 'preferred tokens' - which basically = 'all the fields apiv4 can handle'. Old tokens are 'stuff we also support' - so for example `{contact.gender_id:name}` is  a 'preferred token' - but we don't present it to end users are they likely have no use for it (a better example is `{contribution.contribution_status_id:name}` which is very useful in message templates

eileenmcnaughton replied:
we test `advertisedTokens` so if something causes them to change we decide whether to let the change bubble up or not (e.g someone adds a new field & the token list changes/ test fails - then we decide if the new field should be advertised)

eileenmcnaughton committed:
Add phpdoc to prevent restoreTime being removed

These unused variables cause quite a few bugs - keeping them 'matters' but
they look like they can go and people doing that happens moderately often,
causing bugs (this has been the cause of more than one 'locks don't work bug'

eileenmcnaughton committed:
Superficial test cleanup, protect variable

eileenmcnaughton committed:
Adhere to comment block code standard

Standard is short one liner first:

eileenmcnaughton committed:
Rename variable - I figured out what it meant

eileenmcnaughton committed:
Fix for unresolved state_province

eileenmcnaughton committed:
Add Worldregion api for join

eileenmcnaughton committed:
Fix region token

eileenmcnaughton committed:
Fix preferred_communication method

Also fix over-aggressive test assertion

eileenmcnaughton committed:
Fix key mismatch

eileenmcnaughton committed:
Sigh, and they have colons:

eileenmcnaughton committed:
Style

agileware-justin commented:
Have provided some more details about this issue on https://lab.civicrm.org/dev/core/-/issues/4147#note_88403

totten committed:
Worldregion - For API-entity-naming, prefer "WorldRegion"

totten committed:
Api4 - Rename Worldregion to WorldRegion

eileenmcnaughton commented:
@totten argh that's 15 commits we have to backport too

totten commented:
@eileenmcnaughton Ah, true. But I think git can handle this one time.
